Спринт 9

mezazoy opened the pull request:
…sectionException, TaskNotFoundException.

mezazoy committed:
создать класс BaseHttpHandler, HttpTaskServer, TaskHandler, TaskIntersectionException, TaskNotFoundException.

mezazoy committed:
создать класс SubtaskHandler. Описать в нём методы handle, handleGet, handleGetSubtaskById, handlePost, handleDelete

mezazoy committed:
создать класс EpicHandler. Описать в нём методы handle, handleGet, handleGetSubtaskById, handlePost, handleDelete, handleGetEpicSubtasks

mezazoy committed:
создать класс PrioritizedHandler, HistoryHandler

mezazoy committed:
чекстайл

mezazoy committed:
чекстайл

mezazoy committed:
добавить LocalDateTimeAdapter, DurationTypeAdapter, TaskHandlerTest

mezazoy committed:
чекстайл

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,57 @@
+    private static final int PORT = 8080;
+    private HttpServer server;
+    private final TaskManager taskManager;
+    private static final Gson gson = new GsonBuilder().registerTypeAdapter(Duration.class, new DurationTypeAdapter())
```
Можно было бы вынести  создание  Gson  в базовый  обработчик, в класс BaseHttpHandler. В общем-то, он нужен в обработчиках, а больше особо и нигде. )   

На Ваше усмотрение.

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/DurationTypeAdapter.java:
```diff
@@ -0,0 +1,21 @@
+import java.io.IOException;
+import java.time.Duration;
+
+public class DurationTypeAdapter extends TypeAdapter<Duration> {
```
Самодельные адаптеры принято оформлять как классы и хранить отдельно в пакете, например, adapters

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/EpicHandler.java:
```diff
@@ -0,0 +1,147 @@
+                    handleDelete(exchange);
+                    break;
+                default:
+                    sendNotFound(exchange);
```
default:
                    sendNotFound(exchange);
.--------

 Если пришло непонятно что, то возвращать нужно либо 
 400 BAD_REQUEST
 либо
  405 METHOD_NOT_ALLOWED

Почему возвращается статус 405 METHOD_NOT_ALLOWED можно посмотреть тут: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/EpicHandler.java:
```diff
@@ -0,0 +1,147 @@
+        String path = exchange.getRequestURI().getPath();
+        String[] splitPath = path.split("/");
+        if (splitPath.length < 4) {
+            sendNotFound(exchange);
```
Скорее уж

 400 BAD_REQUEST

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/EpicHandler.java:
```diff
@@ -0,0 +1,147 @@
+
+public class EpicHandler extends BaseHttpHandler implements HttpHandler {
+    private final TaskManager taskManager;
+    private final Gson gson = HttpTaskServer.getGson();
```
private final Gson gson = HttpTaskServer.getGson();

Переменную можно было бы перенести в базовый обработчик, тогда она была бы доступна всем остальным (protected gson)

avfyodorov commented on src/main/java/ru/yandex/taskTraker/server/EpicHandler.java:
```diff
@@ -0,0 +1,147 @@
+import java.util.List;
+
+public class EpicHandler extends BaseHttpHandler implements HttpHandler {
+    private final TaskManager taskManager;
```
Кстати, ссылку на менеджер тоже можно перенести в базовый обработчик, зачем их каждый раз копировать в наследниках?

avfyodorov reviewed (commented):
Добрый день, Марк! 

Хорошая работа, от меня всего пара уточнений. 

Просьба обратить дополнительное внимание на возвращаемые коды ошибок: 
 400 BAD_REQUEST  -  Ошибка в самом запросе,  не удаётся его правильно разобрать. 
 404 NOT FOUND    - это правильный запрос, просто нет данных
 405 METHOD_NOT_ALLOWED  -  Пришло что-то непонятное, с этим мы не работаем 

 https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405

mezazoy committed:
перенести taskManager и gson в базовый обработчик.
добавить возврат корректных кодов ошибок.
создать отдельный пакет для адаптеров LocalDateTimeAdapter и DurationTypeAdapter

avfyodorov reviewed (commented):
Добрый день, Марк!

Замечаний нет.
Работа принята.
